xds/internal/balancer/gracefulswitch/gracefulswitch.go

```diff
@@ -0,0 +1,361 @@
+	// called.
+	bw.Balancer.UpdateSubConnState(sc, state)
+}
+
+// bw must not be referenced via balancerCurrent or balancerPending in gsb
+// before Close is called.
+func (bw *balancerWrapper) Close() {
+	if bw == nil {
+		return
+	}
+	// There is no need to protect this read with a mutex, as Close() is
+	// impossible to be called concurrently with the write in SwitchTo(). The
+	// callsites of Close() for this balancer in Graceful Switch Balancer will
+	// never be called until SwitchTo() returns.
+	bw.Balancer.Close()
+	bw.gsb.mu.Lock()
+	for sc := range bw.subconns {
+		bw.gsb.cc.RemoveSubConn(sc)
+	}
+	bw.gsb.mu.Unlock()
+}
+
+func (bw *balancerWrapper) UpdateState(state balancer.State) {
+	// Hold the mutex for this entire call to ensure it cannot occur
+	// concurrently with other updateState() calls. This causes updates to
+	// lastState and calls to cc.UpdateState to happen atomically.
+	bw.gsb.mu.Lock()
+	defer bw.gsb.mu.Unlock()
+	bw.lastState = state
+
+	if !bw.gsb.balancerCurrentOrPending(bw) {
+		return
+	}
+
+	if bw == bw.gsb.balancerCurrent {
+		// In the case that the current balancer exits READY, and there is a pending
+		// balancer, you can forward the pending balancer's cached State up to
+		// ClientConn and swap the pending into the current. This is because there
+		// is no reason to gracefully switch from and keep using the old policy as
+		// the ClientConn is not connected to any backends.
+		if state.ConnectivityState != connectivity.Ready && bw.gsb.balancerPending != nil {
+			bw.gsb.swap()
+			return
+		}
+		// Even if there is a pending balancer waiting to be gracefully switched to,
+		// continue to forward current balancer updates to the Client Conn. Ignoring
+		// state + picker from the current would cause undefined behavior/cause the
+		// system to behave incorrectly from the current LB policies perspective.
+		// Also, the current LB is still being used by grpc to choose SubConns per
+		// RPC, and thus should use the most updated form of the current balancer.
+		bw.gsb.cc.UpdateState(state)
+		return
+	}
+	// If the current balancer is currently in a state other than READY, the
+	// new policy can be swapped into place immediately. This is because
+	// there is no reason to gracefully switch from and keep using the old
+	// policy as the ClientConn is not connected to any backends.
+	if state.ConnectivityState != connectivity.Connecting || bw.gsb.balancerCurrent.lastState.ConnectivityState != connectivity.Ready {
+		bw.gsb.swap()
+	}
+}
+
+func (bw *balancerWrapper) NewSubConn(addrs []resolver.Address, opts balancer.NewSubConnOptions) (balancer.SubConn, error) {
+	bw.gsb.mu.Lock()
+	if !bw.gsb.balancerCurrentOrPending(bw) {
+		bw.gsb.mu.Unlock()
+		return nil, fmt.Errorf("%T at address %p that called NewSubConn is deleted", bw, bw)
+	}
+	bw.gsb.mu.Unlock()
+
+	sc, err := bw.gsb.cc.NewSubConn(addrs, opts)
+	if err != nil {
+		return nil, err
+	}
+	bw.gsb.mu.Lock()
```

Comment:
Can this balancer be closed while the mu is released? If so, this SubConn may leak.

If this is a problem, we can make the `NewSubConn` first, then hold the mutex and do everything.
Just that if we found this balancer is closed (`balancerCurrentOrPending()` returns false), we need to `RemoveSubConn`.

Reply:
Good catch.  I think checking `balancerCurrentOrPending` _before and after_ the call or holding the lock the whole time would be the way to go.  It wouldn't make sense for long-dead and closed balancers to be able to keep creating subconns which we have to immediately remove.

Reply:
So, what I decided to do is add the conditional of !balancerCurrentOrPending and removal of subconn here instead of merge the original check. This adds code duplication for the same if conditional, but removes unnecessary work for the ClientConn if a NewSubConn() call comes in on an already deprecated (Closed) balancer. Let me know if it's better your suggestion.